homeassistant/components/skybell/__init__.py

```diff
@@ -1,92 +1,166 @@
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): cv.string,
+                    vol.Required(CONF_PASSWORD): cv.string,
+                }
+            )
+        },
+    ),
     extra=vol.ALLOW_EXTRA,
 )
 
+PLATFORMS = [BINARY_SENSOR, CAMERA, LIGHT, SENSOR, SWITCH]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the SkyBell component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    entry_config = {}
+    if DOMAIN not in config:
+        return True
+    for parameter in config[DOMAIN]:
+        if parameter == CONF_USERNAME:
+            entry_config[CONF_EMAIL] = config[DOMAIN][parameter]
+        else:
+            entry_config[parameter] = config[DOMAIN][parameter]
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=entry_config,
+            )
+        )
+
+    return True
+
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Set up the Skybell component."""
-    conf = config[DOMAIN]
-    username = conf.get(CONF_USERNAME)
-    password = conf.get(CONF_PASSWORD)
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Skybell from a config entry."""
+    email = entry.data[CONF_EMAIL]
+    password = entry.data[CONF_PASSWORD]
 
+    api = Skybell(
+        username=email,
+        password=password,
+        get_devices=True,
+        cache_path=hass.config.path(DEFAULT_CACHEDB),
+        session=async_get_clientsession(hass),
+    )
     try:
-        cache = hass.config.path(DEFAULT_CACHEDB)
-        skybell = Skybell(
-            username=username,
-            password=password,
-            get_devices=True,
-            cache_path=cache,
-            agent_identifier=AGENT_IDENTIFIER,
-        )
+        devices = await api.async_initialize()
+    except SkybellAuthenticationException as ex:
+        raise ConfigEntryAuthFailed(
+            f"Authentication Error: please check credentials: {ex}"
+        ) from ex
+    except SkybellException as ex:
+        raise ConfigEntryNotReady(f"Unable to connect to Skybell service: {ex}") from ex
+
+    async def async_update_data() -> None:
+        """Fetch data from API endpoint."""
+        try:
+            [await device.async_update() for device in devices]
+        except SkybellException as err:
+            raise UpdateFailed(f"Failed to communicate with device: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        logging.getLogger(__name__),
+        name=DEFAULT_NAME,
+        update_method=async_update_data,
+        update_interval=MIN_TIME_BETWEEN_UPDATES,
+    )
+    await coordinator.async_config_entry_first_refresh()
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_COORDINATOR: coordinator,
+        DATA_DEVICES: devices,
+    }
```

Comment:
It would be cleaner to make this a dataclass.  Example in `lookin.models`

Reply:
Nice!!!